design/pv_restore_info.md

```diff
@@ -0,0 +1,143 @@
+Two steps are involved in generating the restore volume info, the first is "collection", which is to gather the information 
+for restoration of the volumes, the second is "generation", which is to iterate through the data collected in the first step
+and generate the volume info list as is described above.
+
+Unlike backup, the CR objects created during the restore process will not be persisted to the backup storage location.  
+Therefore, to gather the information needed to generate volume information, we either need to collect the CRs in the middle
+of the restore process, or retrieve the objects based on the `resouce-list.json` of the restore via API server.
+The information to be collected are:
+- **PV/PVC mapping relationship:** It will be collected via the `restore-resource-list.json`, b/c at the time the json is ready, all
+PVCs and PVs are already created.
+- **Native snapshot information:** It will be collected in the restore workflow when each snapshot is restored.
+- **podvolumerestore CRs:** It will be collected in the restore workflow after each pvr is created.
+- **volumesnapshot CRs for CSI snapshot:** It will be collected in the step of collecting PVC info, by reading the `dataSource`
+field in the spec of the PVC.
+- **datadownload CRs** It will be collected in the phase of collecting PVC info, by querying the API-server to list the datadownload
+CRs labeled with the restore name.
+
+After the collection step, the generation step is relatively straight-forward, as we have all the information needed in 
+the data structures.  
+
+The whole collection and generation steps will be done with the "best-effort" manner, i.e. if there are any failures we 
+will only log the error in restore log, rather than failing the whole restore process, we will not put these errors or warnings
+into the `result.json`, b/c it won't impact the restored resources.
+
+Depending on the number of the restored PVCs the "collection" step may involve many API calls, but it's considered acceptable
+b/c at that time the resources are already created, so the actual RTO is not impacted.  By using the client of controller runtime
+we can make the collection step more efficient by using the cache of the API server.  We may consider to make improvements if 
+we observe performance issues, like using multiple go-routines in the collection.
+
+### Implementation
+Because the restore volume info shares the same data structures with the backup volume info, we will refactor the code in 
+package `internal/volume` to make the sub-components in backup volume info shared by both backup and restore volume info.  
+
+We'll introduce a struct called `RestoreVolumeInfoTracker` which encapsulates the logic of collecting and generating the restore volume info:
+```
+// RestoreVolumeInfoTracker is used to track the volume information during restore.
+// It is used to generate the RestoreVolumeInfo array.
+type RestoreVolumeInfoTracker struct {
+	*sync.Mutex
```

Comment:
Could you explain more on why the mutex is needed here?
Didn't find where it is used.

Reply:
It's introduced to make tracker thread-safe, considering velero may improve parallelism in future.
An example of usage:
```
func (t *RestoreVolumeInfoTracker) TrackNativeSnapshot(pvName string, snapshotHandle, volumeType, volumeAZ string, iops int64) {
	t.Lock()
	defer t.Unlock()
	t.pvNativeSnapshotMap[pvName] = NativeSnapshotInfo{
		SnapshotHandle: snapshotHandle,
		VolumeType:     volumeType,
		VolumeAZ:       volumeAZ,
		IOPS:           strconv.FormatInt(iops, 10),
	}
}
```